### DISCUSSION TITLE
XR Implementation Question

### DISCUSSION THREAD

**Discussion by billyjacoby** [callstack/react-native-visionos#105] [Q&A] [answered]
Does the current XR Implementation allow for multiple windows being created and managed from the React Native side, or only the ability to launch a separate SwiftUI container inside the `ImmersiveSpace` closure in the `App.swift` file?

**Comment by okwasniewski** [comment, marked as answer]
Hey! 

Currently it allows to open a separate Swift container with `ImmersiveSpace`. Our goal for this feature is to allow users to be able to render 3D content inside of immersive spaces using JS (Metal) but this requires support of 3rd party modules like Babylon React Native. 

Also an API to open new windows will be added soon 🙌

**Reply by billyjacoby** [reply]
Got it thanks a ton for the reply!

I'm most interested in the API to open new windows. I've got limited native library experience but i'd love to help get this in any way I can!

**Comment by MP185496** [comment]
Is there a way to interact with ImmersiveSpace, like moving the object, adding text above it or changing the text. Right now I saw just opening and closing it in XR? Thanks

**Reply by okwasniewski** [reply]
React Native visionOS provides only core functionality for ImmersiveSpaces. Currently, you can use RealityKit to render UI and 3D elements inside of Immersive Space: https://www.oskarkwasniewski.dev/blog/react-native-visionos-realitykit-snow

We are also planning to add support for BabylonJS in the future

**Reply by billyjacoby** [reply]
Thanks a ton for this post, it's super helpful! I'll be diving more into the Swift part of this stuff as soon as I have a bit more free time.
It would be amazing to at least be able to pass data between these Immersive Views (and eventually WindowGroups) and React Native
